File: 2022/src/day19.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = long long;

const ll MATS = 4;
enum material { ORE = 3, CLAY = 2, OBSIDIAN = 1, GEODE = 0, NONE = 4 };

using mat_array = array<ll, MATS>;
using blueprint = array<mat_array, MATS + 1>;

const map<string, material> material_map = {
    {"ore", ORE},
    {"clay", CLAY},
    {"obsidian", OBSIDIAN},
    {"geode", GEODE},
};

struct solution {
  vector<blueprint> input;

  solution(const string& filename) {
    string line;
    ifstream input_file(filename);
    if (!input_file.is_open()) {
      cerr << "Could not open the file - '" << filename << "'" << endl;
      return;
    }

    regex blueprint_regex("Blueprint (\\d+):");
    regex part_regex("\\s+Each (\\w+) robot costs (\\d+) (\\w+).");
    regex parts_regex(
        "\\s+Each (\\w+) robot costs (\\d+) (\\w+) and (\\d+) (\\w+).");

    ll index = -1;

    while (getline(input_file, line)) {
      smatch match;
      if (regex_match(line, match, blueprint_regex)) {
        index = stoll(match.str(1)) - 1;
        input.resize(index + 1);
      } else if (regex_match(line, match, part_regex)) {
        material mat = material_map.at(match.str(1));
        ll p1_cost = stoll(match.str(2));
        material p1_mat = material_map.at(match.str(3));
        input[index][mat][p1_mat] = p1_cost;
      } else if (regex_match(line, match, parts_regex)) {
        material mat = material_map.at(match.str(1));
        ll p1_cost = stoll(match.str(2));
        material p1_mat = material_map.at(match.str(3));
        ll p2_cost = stoll(match.str(4));
        material p2_mat = material_map.at(match.str(5));
        input[index][mat][p1_mat] = p1_cost;
        input[index][mat][p2_mat] = p2_cost;
      }
    }
    input_file.close();
  }

  void print_blueprint(ll b) {
    printf("blueprint %lld >>\n", b + 1);
    for (ll r = 0; r < MATS; ++r) {
      printf("  r %lld: ", r);
      for (ll i = 0; i < MATS; ++i) {
        printf("%lld, ", input[b][r][i]);
      }
      printf("\n");
    }
  }

  void solve1(bool part2 = false) {
    ll res = part2 ? 1 : 0;
    ll max_time = part2 ? 32 : 24;
    if (part2 && input.size() > 3) {
      input.resize(3);
    }

    for (ll b = 0; b < (ll)input.size(); ++b) {
      mat_array max_cost = {0, 0, 0, 0};
      for (ll r = 0; r < MATS; ++r) {
        for (ll i = 0; i < MATS; ++i) {
          max_cost[i] = max(max_cost[i], input[b][r][i]);
        }
      }

      // dp: time -> robot -> resources
      vector<set<pair<mat_array, mat_array>>> dp(max_time + 1);
      dp[0].insert({{0, 0, 0, 1}, {0, 0, 0, 0}});

      auto update = [&](ll t, material robot, mat_array robots,
                        mat_array resources) {
        for (ll i = 0; i < MATS; ++i) {
          resources[i] += robots[i] - input[b][robot][i];
        }
        if (robot != NONE) {
          robots[robot]++;
        }

        for (ll i = 0; i < MATS; ++i) {
          if (i != GEODE) {
            ll cap = max_cost[i] + (max_time - t) * (max_cost[i] - robots[i]);
            resources[i] = clamp(resources[i], 0ll, cap);
          }
        }
        dp[t].insert({robots, resources});
      };
      for (ll t = 0; t < max_time; ++t) {
        for (const auto& [robots, resources] : dp[t]) {
          for (ll r = 0; r < MATS + 1; ++r) {
            bool ok = r == NONE || r == GEODE || robots[r] < max_cost[r];
            for (ll i = 0; i < MATS; ++i) {
              ok &= resources[i] >= input[b][r][i];
            }
            if (!ok) {
              continue;
            }

            update(t + 1, (material)r, robots, resources);
            if (r == GEODE) {  // should always build geode when possible?
              break;
            }
          }
        }

        // printf("  t %lld, next size: %ld\n", t + 1, dp[t + 1].size());
        dp[t] = {};
      }

      ll max_geode = 0;
      for (const auto& [_, resources] : dp[max_time]) {
        max_geode = max(max_geode, resources[GEODE]);
      }
      if (part2) {
        res *= max_geode;
      } else {
        res += (b + 1) * max_geode;
      }
      printf("blueprint %lld max_geodes: %lld\n", b + 1, max_geode);
    }

    printf("res: %lld\n", res);
  }

  void solve2() { solve1(true); }
};

int main(int argc, char** argv) {
  const string day = argc > 1 ? argv[1] : "";
  const bool final = argc > 2 && string(argv[2]) == "1";
  const bool snd_question = argc > 3 && string(argv[3]) == "1";
  const string filename =
      "./input/" + string(final ? "final" : "practice") + "/day" + day + ".txt";
  snd_question ? solution(filename).solve2() : solution(filename).solve1();
}

File: 2022/rust/src/bin/day19.rs
use std::collections::HashMap;

use nom::{
    bytes::complete as nbc, character::complete as ncc, combinator as nc, multi as nm,
    sequence as ns, IResult,
};

use aoc_2022::*;

#[derive(Debug)]
struct Data {
    data: Vec<[Mats; Mat::MAT_COUNT]>,
    ids: Vec<usize>,
}

impl Day for Data {
    type Output = usize;

    fn parse(input: &str) -> IResult<&str, Self> {
        let id = |i| ns::delimited(nbc::tag("Blueprint "), ncc::u64, nbc::tag(":"))(i);

        let material =
            |i| ns::separated_pair(ncc::u64, nbc::tag(" "), nc::map(ncc::alpha1, Mat::new))(i);
        let robot = |i| {
            nc::map(
                ns::separated_pair(
                    nc::map(ns::preceded(nbc::tag("Each "), ncc::alpha1), Mat::new),
                    nbc::tag(" robot costs "),
                    ns::terminated(
                        nm::separated_list1(nbc::tag(" and "), material),
                        nbc::tag("."),
                    ),
                ),
                |(mat, l)| {
                    let mut costs = [0; Mat::MAT_COUNT];
                    for (count, mat) in l {
                        costs[mat as usize] = count as usize;
                    }
                    (mat, costs)
                },
            )(i)
        };

        let blueprint = |i| {
            nc::map(
                ns::pair(
                    whitespace(id),
                    nm::many_m_n(Mat::MAT_COUNT, Mat::MAT_COUNT, whitespace(robot)),
                ),
                |(id, l)| {
                    let mut robot = [[0; Mat::MAT_COUNT]; Mat::MAT_COUNT];
                    for (mat, costs) in l {
                        robot[mat as usize] = costs;
                    }
                    (id as usize, robot)
                },
            )(i)
        };

        let result = nc::map(nm::many0(blueprint), |info| {
            let (ids, data) = info.into_iter().unzip();
            Self { data, ids }
        })(input);
        result
    }

    fn part1(&mut self) -> Self::Output {
        self.data
            .iter()
            .map(|blueprint| Dfs::new(blueprint, 24).solve())
            .enumerate()
            .map(|(blueprint, max_geodes)| self.ids[blueprint] * max_geodes)
            .sum()
    }

    fn part2(&mut self) -> Self::Output {
        self.data
            .iter()
            .take(3)
            .map(|blueprint| Dfs::new(blueprint, 32).solve())
            .product()
    }
}

#[derive(Debug)]
#[repr(usize)]
enum Mat {
    Ore,
    Clay,
    Obsidian,
    Geode,
}

impl Mat {
    const MAT_COUNT: usize = 4;

    fn new(s: &str) -> Self {
        match s {
            "ore" => Self::Ore,
            "clay" => Self::Clay,
            "obsidian" => Self::Obsidian,
            "geode" => Self::Geode,
            _ => panic!("invalid material"),
        }
    }
}

type Mats = [usize; Mat::MAT_COUNT];

struct Dfs<'a> {
    blueprint: &'a [Mats; Mat::MAT_COUNT],
    max_time: usize,
    max_cost: Mats,
    max_geodes: usize,
    memo: HashMap<(Mats, Mats), usize>,
}
impl<'a> Dfs<'a> {
    fn new(blueprint: &'a [Mats; Mat::MAT_COUNT], max_time: usize) -> Self {
        let mut max_cost = [0; Mat::MAT_COUNT];
        for mat in 0..Mat::MAT_COUNT {
            let max_opt = blueprint.iter().map(|robot| robot[mat]).max();
            max_cost[mat] = if mat == Mat::Geode as usize {
                (max_time - 1) * (max_time) / 2
            } else {
                max_opt.unwrap_or_default()
            }
        }

        Self {
            blueprint,
            max_time,
            max_cost,
            max_geodes: 0,
            memo: HashMap::new(),
        }
    }

    fn solve(&mut self) -> usize {
        let mut robots = [0; Mat::MAT_COUNT];
        robots[Mat::Ore as usize] = 1;
        let resources = [0; Mat::MAT_COUNT];

        self.dfs(self.max_time, robots, resources);
        self.max_geodes
    }

    fn can_add(&self, robot: usize, resources: &Mats) -> bool {
        (0..Mat::MAT_COUNT).all(|mat| resources[mat] >= self.blueprint[robot][mat])
    }

    fn dfs(&mut self, time: usize, robots: Mats, resources: Mats) {
        self.memo.insert((robots, resources), time);
        self.max_geodes = resources[Mat::Geode as usize].max(self.max_geodes);
        if time == 0 {
            return;
        }

        for robot in 0..Mat::MAT_COUNT {
            let mut time_left = time - 1;
            let mut next_resources = resources;
            let mut next_robots = robots;

            while time_left != 0 && !self.can_add(robot, &next_resources) {
                for mat in 0..Mat::MAT_COUNT {
                    next_resources[mat] += robots[mat];
                }
                time_left -= 1;
            }

            let add = self.can_add(robot, &next_resources);
            if add {
                next_robots[robot] = (robots[robot] + 1).min(self.max_cost[robot]);
            }

            for mat in 0..Mat::MAT_COUNT {
                let unmet_count = self.max_cost[mat] - next_robots[mat];
                let cap = self.max_cost[mat] + time_left * unmet_count;
                let cost = if add { self.blueprint[robot][mat] } else { 0 };
                next_resources[mat] = (next_resources[mat] + robots[mat] - cost).min(cap);
            }

            let potential_geodes = next_resources[Mat::Geode as usize]
                + time_left * next_robots[Mat::Geode as usize]
                + (time_left - 1) * (time_left) / 2;
            if potential_geodes > self.max_geodes {
                let better_time = (self.memo.get(&(next_robots, next_resources)))
                    .map(|&t| t >= time_left)
                    .unwrap_or_default();
                if !better_time {
                    self.dfs(time_left, next_robots, next_resources);
                }
            }
        }
    }
}

fn main() {
    Data::solve();
}
